### PR TITLE
add s3 interface

### PR DESCRIPTION
**Pull request by delgod**

### CODE REVIEW

File: interfaces/s3/v0/schemas/provider.json [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+            "title": "S3 URI Style",
+            "description": "The S3 protocol specific bucket path lookup type.",
+            "type": "string",
+            "default": "",
+            "examples": [
+                "path",
+                "host"
+            ]
+        },
+        "storage-class": {
```
[the remark this answers is not in the document]
Comment (ChrisMacNaughton) [reply interfaces/s3/v0/schemas/provider.json:79]: Does that mean that `storage-class` is a client parameter used to request storage class? ie: maybe it should return a list of available storage classes?

**Comment by ChrisMacNaughton** [comment]
It also occurred to me this weekend that we use boto3 as an S3 client to do some testing of RadosGW already, and it looks like it all just works with just endpoint_url, access_key, and secret_key so I'm fairly optimistic that the interface will work fairly well with it as well

File: interfaces/s3/v0/schemas/provider.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (delgod) [reply interfaces/s3/v0/schemas/provider.json:79]: yes, `storage-class` is a client parameter e.g. [boto3 supports it in `put_object` call](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/s3.html#S3.Client.put_object).
But the list of available storage classes is a vendor-specific thing. So we cannot use enum here (on interface level).
It this PR, we discuss which information is passed from the S3 service to the S3 client.
I don't think that we should return all possible storage classes here, only one class should be returned.

**Comment by ChrisMacNaughton** [comment]
I think that adding in storage-class to the interface is asking for an explosion of (nearly) duplicate relations to handle giving S3 clients the storage-class an admin prefers. Digging through the S3 documentation about storage-classes, and comparing it to Ceph RadosGW, it looks like we would be supporting _at least_ 8 different classes of storage / nearly identical relations.

I'd lean towards a config option on charms consuming S3 to specify desired storage-class paired with a reasonable default (S3 and RadosGW seen to both default to `STANDARD` so that seems like a reasonable choice). Additionally, this relation could include something like `supported-storage-classes` so that, for example, a RadosGW could include only the storage classes that are supported in this installation.

Looking back at previous discussions here about storage-classes, the Ceph docs around them say that a user should only use the storage-class names as defined by Amazon, as the common libraries do things like raise exceptions and ignore the call if unknown storage-classes are used.

File: interfaces/s3/v0/README.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+- It is expected to provide an optional `storage-class` field for the S3 storage class.
+- It is expected to provide an optional `tls-ca-chain` field for TLS verification. The field can take a list of strings. Each string should be in base64 form and represent one certificate. All certificates together should represent a complete CA chain which can be used for HTTPS validation.
+- It is expected to provide an optional `s3-api-version` field for the (S3 protocol specific) API signature. The field can take only `2` and `4` values.
+- It is expected to provide an optional `attributes` field for the custom metadata. The field can take a list of strings. Server-Side-Encryption headers should be passed into this field, if any.
+
+### Requirer
+- Is expected to provide a bucket name in the `bucket` field. Field value should be generated on Requirer side if no particular value set in Requirer juju config.
+- Is expected than Provider can ignore `bucket` field in some cases (e.g. S3Proxy or S3 Integrator) and provide correct bucket name back.
```
Comment (simskij) [inline interfaces/s3/v0/README.md:34]:
```suggestion
- Is expected to tolerate that the Provider may ignore the `bucket` field in some cases (e.g. S3Proxy or S3 Integrator) and instead use the bucket name received.
```

**Comment by simskij** [comment]
@delgod and I just finished a design session on this, and came to the following conclusion:

There are two pragmatic ways of solving this. Either we have multiple named relations for the same interface (`s3`) in the provider end or we have them in the consumer end. 

While both of these would potentially work, the latter would mean that each consumer of the s3 interface would have to know all the storage class names each provider implements and keep a mapping for which to use and when, shifting unnecessarily large amounts of knowledge about the underlying storage provider either to a) the admin, or b) the consuming charm dev.

On the other hand, if we have the provider worry about adding one named relation with the same interface (still `s3`) per storage class they support, they would then be able to use different event subscriptions to serve buckets with different storage classes.

Setting the storage class at runtime through a config option is not really an option, in our opinion, as it yet again shifts the responsibility to the admin of knowing what type of storage class to configure for each use case. For example, doing delta online backups of a DBMS has vastly different requirements than doing a yearly full backup of data with legal retention requirements.

@ChrisMacNaughton, would this solution be acceptable? For instance, for the radosgw charm to provide one named relation per storage class using the same `s3` interface? If not, I think we might need to book a session with the three of us together to try to iron this out as it currently is a blocker for @delgod's team. 😅

File: interfaces/s3/v0/README.md (hunk 2 above)
Comment (simskij) [reply interfaces/s3/v0/README.md:34]: Apparently @delgod, i'm not allowed to press "Commit change" on this suggestion as I don't have write access to your fork 😅

**Comment by jnsgruk** [comment]
> @delgod and I just finished a design session on this, and came to the following conclusion:
> 
> There are two pragmatic ways of solving this. Either we have multiple named relations for the same interface (`s3`) in the provider end or we have them in the consumer end.
> 
> While both of these would potentially work, the latter would mean that each consumer of the s3 interface would have to know all the storage class names each provider implements and keep a mapping for which to use and when, shifting unnecessarily large amounts of knowledge about the underlying storage provider either to a) the admin, or b) the consuming charm dev.
> 
> On the other hand, if we have the provider worry about adding one named relation with the same interface (still `s3`) per storage class they support, they would then be able to use different event subscriptions to serve buckets with different storage classes.
> 
> Setting the storage class at runtime through a config option is not really an option, in our opinion, as it yet again shifts the responsibility to the admin of knowing what type of storage class to configure for each use case. For example, doing delta online backups of a DBMS has vastly different requirements than doing a yearly full backup of data with legal retention requirements.
> 
> @ChrisMacNaughton, would this solution be acceptable? For instance, for the radosgw charm to provide one named relation per storage class using the same `s3` interface? If not, I think we might need to book a session with the three of us together to try to iron this out as it currently is a blocker for @delgod's team. sweat_smile

FTR I think this is a good solution - it shifts the burden of deciding which storage class is most suitable to the charm developer, who should ultimately be the expert in the application.

+1 for one single interface, with storage class being specified by the charm in the constructor call for the lib. If multiple storage classes are applicable, multiple endpoints should be declared.

**Comment by ChrisMacNaughton** [comment]
taking my above example a bit further handling a case where multiple of S3 are desired:

```bash
# config based, using defaults (backup_dataclass=standard, archive_dataclass=glacier)
juju deploy mysql

# config based
juju deploy mysql --config backup_dataclass=standard-ia --config archive_dataclass=outpost


# add the single relation to s3
juju add-relation mysql s3_provider
# we might want different relations
juju add-relation mysql:backup s3_provider
juju add-relation mysql:archive s3_provider
```

```bash
# no need for config
juju deploy mysql
# must specify remote relation for each backend, cannot reuse an s3 interface
juju add-relation mysql:backup s3_provider:standard
juju add-relation mysql:archive s3_provider:glacier
# and what if a charm backend named the data-type differently?
# go read through that charm's docs to figure out what's available
```

**Comment by jnsgruk** [comment]
> To resolve the "know what each storage backend supports", `storage-class` could be replaced with `available-storage-classes`, and potentially `default-storage-class` to cover the case where a user-requested storage-class isn't available.
> 
> > Setting the storage class at runtime through a config option is not really an option, in our opinion, as it yet again shifts the responsibility to the admin of knowing what type of storage class to configure for each use case. For example, doing delta online backups of a DBMS has vastly different requirements than doing a yearly full backup of data with legal retention requirements.
> 
> In this example, an admin would still have to choose storage class at deployment, ie: it's a comparison between these two uses rather than _magic_:
> 
> ```
> # using config, and in the example backup_dataclass should have a default as well
> juju deploy mysql --config backup_dataclass=mycustom
> juju add-relation mysql s3_provider
> # using  relations, defaults are impossible to set here
> juju deploy mysql
> juju add-relation mysql:s3 s3_provider:mycustom
> ```
> 
> For a storage provider that _can_ have dynamic (and user defined!) storage classes, relation based config cannot work, as relations are defined statically at charm-build time.

I think my biggest problem here is with the use of config on the MySQL itself. Take the S3 Integrator as an example, it will be in the model representing a bucket, essentially. In that case, it makes no sense to configure MySQL for a storage class at all - you'd configure the S3 integrator to hand over the storage class of the thing its representing 

That might be a poor explanation, but hopefully that clarifies my thinking a little.

**Comment by ChrisMacNaughton** [comment]
Using a more dynamic example (Ceph RadosGW), an operator isn't restricted in what storage-classes they could create. In a relation driven flow, the operator must only use our prescribed approach to configure their RadosGW or find themselves unable to configure related units to use their desired storage pools. Looking through the documentation, storage-class isn't a bucket config either, it's an object config (which is why we're passing it to the client in the first place!).

I feel like the best approach to this would be something like a relation config option, where an admin could set, on the s3 relation, a config option to select the storage class for this instance of this relation.

To maximise compatibility and success likelihood, I think it would work for the ceph-radosgw charm to only support the default storage class (standard) as it is always created by Ceph, at which point the complexity around relations becomes a bit moot for Ceph RadosGW, and an operator could customise that class in Ceph if it were deemed necessary.

**Comment by jnsgruk** [comment]
Yeh I think that's fair. We could easily go with the following behaviour:

- the s3 interface just always uses whatever the standard storage class is
- there is no control over that from either the providing or requiring charm

We would then need an `s3-storage-class-config` charm, or similar, much like the `traefik-route-k8s` charm, which would allow the operator to deploy it in between the provider and consumer, specifying their required storage class as config? (This is essentially "relation config", and can be replaced with a unit-less charm in months to come...)

Thoughts on that?

**Comment by ChrisMacNaughton** [comment]
As much as I'm not a fan of that type of 'relation config via charm', I do think it sounds like it more cleanly models the concept; both allowing for easy, reasonable defaults while enabling customisation when needed

**Comment by simskij** [comment]
> > As much as I'm not a fan of that type of 'relation config via charm', I do think it sounds like it more cleanly models the concept; both allowing for easy, reasonable defaults while enabling customisation when needed
> 
> Yeh, it's a bit of a blunt instrument. FWIW we're gonna make that all much slicker and less invasive once we introduce scriptlets to charms. We should talk about that in Prague, perhaps.
> 
> For now: @delgod / @simskij - we good on this proposal? Probably only needs minor changes here to represent.

I'm fine with that, although it means someone will need to write a relation interface for `s3-config-proxy` or whatever we would call it as well. Any takers?

Are you fine with the proposal, @delgod? Basically, if I understood it correctly:

```plain

  MySQL <- (s3) -> S3 Config Charm <- (s3-config) -> S3 Storage Charm

```

for instance:

```plain

  mysql-k8s:s3-backup > (s3) > s3-config-k8s > (s3-config) > radosgw
           :s3-cold   > (s3) > s3-config-k8s > (s3-config) > aws

```

and so on.

**Comment by simskij** [comment]
image

**Comment by ChrisMacNaughton** [comment]
One more thought about this interface: is there any reason we can't do the s3 interface on both sides of the config charm, ie: rather than this:

```
mysql-k8s:s3-backup > (s3) > s3-config-k8s > (s3-config) > radosgw
         :s3-cold   > (s3) > s3-config-k8s > (s3-config) > aws
```

We could do:

```
mysql-k8s:s3-backup > (s3) > s3-config-k8s > (s3) > radosgw
         :s3-cold   > (s3) > s3-config-k8s > (s3) > aws
```

The only implication for this that I can think of from an interface perspective would be that the storage-class would have to be optional in the interface, and a consuming charm must be able to accept it being unset (or we always set it to the default storage-class in the provider)

**Comment by jnsgruk** [comment]
Yeh, I don't think that's a bad idea. I think we'd already landed on making the field optional. It would save us from implementing another interface :)

**Comment by simskij** [comment]
@delgod, please update this pr to include a consumer schema that optionally requests a storage class; then we're good to merge this.

**Pull request merged**